src/axom/core/ArrayBase.hpp

```diff
@@ -728,20 +732,18 @@ struct ArrayOpsBase;
 template <typename T>
 struct ArrayOpsBase<T, false>
 {
-  static constexpr bool DefaultCtor = std::is_default_constructible<T>::value;
+  using DefaultCtorTag = std::is_default_constructible<T>;
 
   /*!
-   * \brief Default-initializes the "new" segment of an array
+   * \brief Helper for default-initializing the "new" segment of an array
    *
    * \param [inout] data The data to initialize
    * \param [in] begin The beginning of the subset of \a data that should be initialized
    * \param [in] end The end index (exclusive) of the subset of \a data that is to be initialized
    * \note Specialization for when T is default-constructible.
```

Comment:
Suggestion: It might be helpful to add a doxygen `tparam` for `U` and, e.g. how it is related to the note about default-constructibility (here and elsewhere)

Reply:
Those dummy template parameters were originally intended to avoid instantiating the helpers with non-matching tags; however, apparently we get "lazy" instantiation of member functions for free on a template class.

Reply:
(Based on experience) it's worth testing that conjecture with msvc... it can sometimes exhibit different behavior than other compilers on this sort of thing

Reply:
We don't have a unit test for this, but the `nontrivial_ctor_obj` test in `core_array_for_all` should exercise the non-trivially constructible path for both host and device memory. (we also have tests for nontrivial destructor and copy constructor)

I think we could consider adding a test case for trivial type initialization (int or double), and a test case for non-trivally default-constructible type.